client/files.go

```diff
@@ -0,0 +1,127 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+)
+
+// MakeDirOptions holds the options for a call to MakeDir.
+type MakeDirOptions struct {
+	// Path is the absolute path of the directory to be created (required).
+	Path string
+
+	// MakeParents specifies whether the non-existing parent directories will
+	// be created or not. If one of the parent directories of the specified
+	// path does not exist and MakeParents is false, the call will fail.
+
+	// MakeParents, if true, specifies that any non-existent parent directories
+	// should be created. If false (the default), the call will fail if the
+	// directory to be created has at least one parent directory that does not
+	// exist.
+	MakeParents bool
+
+	// Permissions specifies the permission bits of the directories to be created.
+	Permissions *os.FileMode
+
+	// UserID indicates the ID of the owner user for the created directories.
+	UserID *int
+
+	// User indicates the name of the owner user for the created directories.
+	// If used together with UserID, this value must match the name of the user
+	// with that ID.
+	User string
+
+	// GroupID indicates the ID of the owner group for the created directories.
+	GroupID *int
+
+	// Group indicates the name of the owner group for the created directories.
+	// If used together with GroupID, this value must match the name of the group
+	// with that ID.
+	Group string
+}
+
+type makeDirPayload struct {
+	Action string         `json:"action"`
+	Dirs   []makeDirsItem `json:"dirs"`
+}
+
+type makeDirsItem struct {
+	Path        string `json:"path"`
+	MakeParents bool   `json:"make-parents"`
+	Permissions string `json:"permissions"`
+	UserID      *int   `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     *int   `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+type fileResult struct {
+	Path  string       `json:"path"`
+	Error *errorResult `json:"error,omitempty"`
+}
+
+type errorResult struct {
+	Message string `json:"message"`
+}
+
+// MakeDir creates a directory or directory tree.
+func (client *Client) MakeDir(opts *MakeDirOptions) error {
+	var permissions string
+	if opts.Permissions != nil {
+		permissions = fmt.Sprintf("%o", *opts.Permissions)
+	}
+
+	payload := &makeDirPayload{
+		Action: "make-dirs",
+		Dirs: []makeDirsItem{
+			{
+				Path:        opts.Path,
+				MakeParents: opts.MakeParents,
+				Permissions: permissions,
+				UserID:      opts.UserID,
+				User:        opts.User,
+				GroupID:     opts.GroupID,
+				Group:       opts.Group,
+			},
+		},
+	}
+
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return err
+	}
+
+	var result []fileResult
+	_, err = client.doSync("POST", "/v1/files", nil, nil, &body, &result)
```

Comment:
This currently fails against a real Pebble with the error `invalid Content-Type ""`. We need to add a `Content-Type` header:

```diff
$ git diff
diff --git a/client/files.go b/client/files.go
index 1404dce..45bf8f9 100644
--- a/client/files.go
+++ b/client/files.go
@@ -110,7 +110,10 @@ func (client *Client) MakeDir(opts *MakeDirOptions) error {
        }
 
        var result []fileResult
-       _, err = client.doSync("POST", "/v1/files", nil, nil, &body, &result)
+       headers := map[string]string{
+               "Content-Type": "application/json",
+       }
+       _, err = client.doSync("POST", "/v1/files", nil, headers, &body, &result)
        if err != nil {
                return err
        }
```

To catch issues like this, it would also be good if we had at least one client-to-real-server test to test this (same with the other file APIs).

I tested locally by running `pebble run` in one terminal window, then `pebble mkdir` in another.